x-pack/plugin/search-business-rules/src/main/java/org/elasticsearch/xpack/searchbusinessrules/RuleQueryBuilder.java

```diff
@@ -0,0 +1,361 @@
+    protected QueryBuilder doRewrite(QueryRewriteContext queryRewriteContext) throws IOException {
+        if (curatedIds.isEmpty() == false) {
+            return this;
+        } else if (curatedIdSupplier != null) {
+            List<String> curatedIds = curatedIdSupplier.get();
+            if (curatedIds == null) {
+                return this; // not executed yet
+            } else {
+                return new RuleQueryBuilder(organicQuery, matchCriteria, rulesetIds, curatedIds);
+            }
+        }
+
+        // This is a POC, when we have a query rules CRUD API we'd be calling that instead of this test search.
+        SearchRequest searchRequest = new SearchRequest("demo-curations");
+        searchRequest.preference("_local");
+        searchRequest.source().query(buildCurationQuery());
+        searchRequest.source(new SearchSourceBuilder().query(buildCurationQuery()));
+
+        SetOnce<List<String>> idSetOnce = new SetOnce<>();
+        queryRewriteContext.registerAsyncAction((client, listener) -> {
+            client.search(searchRequest, ActionListener.wrap(response -> {
+                List<String> ids = new ArrayList<>();
+                for (SearchHit hit : response.getHits().getHits()) {
+                    // No error case handling here for POC
+                    Object actions = Objects.requireNonNull(hit.getSourceAsMap()).get("actions");
+                    List<String> idsArray = ((List<?>) actions).stream()
+                        .map(action -> (Map<?, ?>) action)
+                        .map(actionMap -> actionMap.get("ids"))
+                        .flatMap(_ids -> ((List<?>) _ids).stream())
+                        .map(id -> (String) id)
+                        .toList();
+                    ids.addAll(idsArray);
+                }
+                idSetOnce.set(ids.stream().distinct().toList());
+                listener.onResponse(null);
+            }, listener::onFailure));
+        });
+
+        QueryBuilder newOrganicQuery = organicQuery.rewrite(queryRewriteContext);
+        RuleQueryBuilder rewritten = new RuleQueryBuilder(newOrganicQuery, matchCriteria, rulesetIds, idSetOnce::get);
+        rewritten.boost(this.boost);
+        return rewritten;
+    }
+
+    private QueryBuilder buildCurationQuery() {
+        BoolQueryBuilder boolQueryBuilder = QueryBuilders.boolQuery();
+        boolQueryBuilder.filter(QueryBuilders.termsQuery("ruleset_id", rulesetIds));
+        boolQueryBuilder.must(QueryBuilders.termQuery("rule_type", "pinned"));
+        for (MatchCriteria criteria : matchCriteria) {
+            BoolQueryBuilder boolSubQueryBuilder = QueryBuilders.boolQuery();
+            boolSubQueryBuilder.must(QueryBuilders.termQuery("criteria.name", "query_string_match"));
+            boolSubQueryBuilder.must(QueryBuilders.termQuery("criteria.value", criteria.queryStringMatch()));
+            boolQueryBuilder.should(boolSubQueryBuilder);
+        }
+        return boolQueryBuilder;
```

Comment:
Interesting that you took this path. If we're making it a full search thing then I'd recommend to use the percolator. 
Although I am not sure that this is the right tradeoff, normal search or percolator included.
Not very important for this POC either so let's leave this for later. I'll focus on the integration of the idea in the query builder.